File: examples/ai-chat/android/app/src/main/java/ai/connct_screen/com/LlmClient.java
package ai.connct_screen.com;

import android.util.Log;

import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.util.List;
import java.util.concurrent.TimeUnit;

import okhttp3.MediaType;
import okhttp3.OkHttpClient;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;

public class LlmClient {

    private static final String TAG = "A11yAgent";
    private static final MediaType JSON = MediaType.get("application/json; charset=utf-8");
    private static final int MAX_RETRIES = 2;
    private static final long RETRY_DELAY_MS = 2000;

    private final OkHttpClient client;
    private final String apiUrl;
    private final String apiKey;
    private final String model;

    public LlmClient(String apiUrl, String apiKey, String model) {
        this.apiUrl = apiUrl;
        this.apiKey = apiKey;
        this.model = model;
        this.client = new OkHttpClient.Builder()
                .connectTimeout(30, TimeUnit.SECONDS)
                .readTimeout(120, TimeUnit.SECONDS)
                .writeTimeout(30, TimeUnit.SECONDS)
                .build();
    }

    public static class LlmResponse {
        public final String content;
        public final JSONArray toolCalls; // null if no tool calls

        public LlmResponse(String content, JSONArray toolCalls) {
            this.content = content;
            this.toolCalls = toolCalls;
        }

        public boolean hasToolCalls() {
            return toolCalls != null && toolCalls.length() > 0;
        }
    }

    private boolean isRetryable(int httpCode) {
        return httpCode == 429 || httpCode >= 500;
    }

    public LlmResponse chat(List<JSONObject> messages, JSONArray tools) throws IOException, JSONException {
        JSONObject requestBody = new JSONObject();
        requestBody.put("model", model);

        JSONArray messagesArray = new JSONArray();
        for (JSONObject msg : messages) {
            messagesArray.put(msg);
        }
        requestBody.put("messages", messagesArray);

        if (tools != null && tools.length() > 0) {
            requestBody.put("tools", tools);
        }

        String bodyStr = requestBody.toString();
        Log.d(TAG, "[LLM] Request body length: " + bodyStr.length());

        IOException lastException = null;

        for (int attempt = 0; attempt <= MAX_RETRIES; attempt++) {
            if (attempt > 0) {
                Log.d(TAG, "[LLM] Retry attempt " + attempt + "/" + MAX_RETRIES);
                try {
                    Thread.sleep(RETRY_DELAY_MS);
                } catch (InterruptedException e) {
                    Thread.currentThread().interrupt();
                    throw new IOException("Interrupted during retry delay", lastException);
                }
            }

            RequestBody body = RequestBody.create(bodyStr, JSON);
            Request request = new Request.Builder()
                    .url(apiUrl)
                    .addHeader("Authorization", "Bearer " + apiKey)
                    .addHeader("Content-Type", "application/json")
                    .post(body)
                    .build();

            try (Response response = client.newCall(request).execute()) {
                if (!response.isSuccessful()) {
                    String errorBody = response.body() != null ? response.body().string() : "no body";
                    lastException = new IOException("LLM API error " + response.code() + ": " + errorBody);
                    if (isRetryable(response.code()) && attempt < MAX_RETRIES) {
                        Log.d(TAG, "[LLM] Retryable error " + response.code() + ", will retry");
                        continue;
                    }
                    throw lastException;
                }

                String responseStr = response.body().string();
                JSONObject responseJson = new JSONObject(responseStr);
                JSONObject choice = responseJson.getJSONArray("choices").getJSONObject(0);
                JSONObject message = choice.getJSONObject("message");

                String content = message.optString("content", null);
                JSONArray toolCalls = message.optJSONArray("tool_calls");

                return new LlmResponse(content, toolCalls);
            } catch (IOException e) {
                lastException = e;
                if (attempt < MAX_RETRIES) {
                    Log.d(TAG, "[LLM] IOException: " + e.getMessage() + ", will retry");
                    continue;
                }
                throw e;
            }
        }

        // Should not reach here, but just in case
        throw lastException != null ? lastException : new IOException("LLM request failed after retries");
    }
}

File: examples/rn/src/LlmClient.ts
import type { ChatMessage, ToolCall, ToolDefinition } from "./types";

const MAX_RETRIES = 2;
const RETRY_DELAY_MS = 2000;

export interface LlmResponse {
  content: string | null;
  toolCalls: ToolCall[] | null;
}

export class LlmClient {
  private apiUrl: string;
  private apiKey: string;
  private model: string;

  constructor(apiUrl: string, apiKey: string, model: string) {
    this.apiUrl = apiUrl;
    this.apiKey = apiKey;
    this.model = model;
  }

  async chat(
    messages: ChatMessage[],
    tools: ToolDefinition[],
    signal?: AbortSignal
  ): Promise<LlmResponse> {
    const body: Record<string, unknown> = {
      model: this.model,
      messages
    };
    if (tools.length > 0) {
      body.tools = tools;
    }

    const bodyStr = JSON.stringify(body);
    let lastError: Error | null = null;

    for (let attempt = 0; attempt <= MAX_RETRIES; attempt++) {
      if (attempt > 0) {
        await new Promise((r) => setTimeout(r, RETRY_DELAY_MS));
      }

      try {
        const response = await fetch(this.apiUrl, {
          method: "POST",
          headers: {
            Authorization: `Bearer ${this.apiKey}`,
            "Content-Type": "application/json"
          },
          body: bodyStr,
          signal
        });

        if (!response.ok) {
          const errorBody = await response.text();
          lastError = new Error(
            `LLM API error ${response.status}: ${errorBody}`
          );
          if (
            (response.status === 429 || response.status >= 500) &&
            attempt < MAX_RETRIES
          ) {
            continue;
          }
          throw lastError;
        }

        const data = await response.json();
        const choice = data.choices[0];
        const message = choice.message;

        return {
          content: message.content ?? null,
          toolCalls:
            message.tool_calls && message.tool_calls.length > 0
              ? message.tool_calls
              : null
        };
      } catch (e: any) {
        if (e.name === "AbortError") throw e;
        lastError = e;
        if (attempt < MAX_RETRIES) continue;
        throw e;
      }
    }

    throw lastError ?? new Error("LLM request failed after retries");
  }
}
